OpenRA.Mods.Common/AI/HackyAI.cs

```diff
@@ -231,6 +238,30 @@ public void Activate(Player p)
 				resourceTypeIndices.Set(World.TileSet.GetTerrainIndex(t.TerrainType), true);
 		}
 
+		// TODO: Give this a more generic name
+		public bool EnoughWaterToBuildNaval()
+		{
+			var baseBuildings = World.Actors.Where(
+				a => a.Owner == Player
+					&& a.HasTrait<BaseBuilding>()
+					&& !a.HasTrait<Mobile>());
+
+			foreach (var b in baseBuildings)
+			{
+				// TODO: Unhardcode terrain type
+				// TODO2: Only count cells that have all 8 adjacent cells free and of matching terrain type, then anything > 0 would be fine
+				var countWaterTiles = Map.FindTilesInCircle(b.Location, Info.MaxBaseRadius)
+					.Where(c => Player.World.Map.Contains(c)
+						&& Player.World.Map.GetTerrainInfo(c).Type == "Water")
+					.Count();
```

Comment:
Shouldn't this check if the water is in buildable range, too (`BuildingInfo.Adjacent`)? Otherwise the AI would still try to build naval structures even before it could actually place them just because there is water somewhere within the base radius. Not sure, really. Maybe this is good enough.